In [1]:
import warnings
warnings.filterwarnings('ignore')
import pandas as pd
import math
from sklearn.model_selection import train_test_split,KFold
import numpy as np

In [2]:
datafile_train=r'C:\Users\Anmol\Desktop\Dig_Nest_Data\Python_Data_Materials\Main Project\P1 Data\Consumer_Complaints_train.csv'
datafile_test=r'C:\Users\Anmol\Desktop\Dig_Nest_Data\Python_Data_Materials\Main Project\P1 Data\Consumer_Complaints_test_share.csv'
cd_train=pd.read_csv(datafile_train)
cd_test=pd.read_csv(datafile_test)

In [3]:
cd_train.dtypes

Date received                   object
Product                         object
Sub-product                     object
Issue                           object
Sub-issue                       object
Consumer complaint narrative    object
Company public response         object
Company                         object
State                           object
ZIP code                        object
Tags                            object
Consumer consent provided?      object
Submitted via                   object
Date sent to company            object
Company response to consumer    object
Timely response?                object
Consumer disputed?              object
Complaint ID                     int64
dtype: object

In [4]:
for col in ['Date received','Date sent to company']:
    cd_train[col]=pd.to_datetime(cd_train[col],infer_datetime_format=True)
    cd_test[col]=pd.to_datetime(cd_test[col],infer_datetime_format=True)

In [5]:
cd_train['day_diff']=pd.to_numeric(cd_train['Date sent to company']-cd_train['Date received'])
cd_test['day_diff']=pd.to_numeric(cd_test['Date sent to company']-cd_test['Date received'])/86400000000000

In [6]:
for col in ['Date received','Date sent to company']:
    cd_train.drop([col],1,inplace=True)
    cd_test.drop([col],1,inplace=True)

In [7]:
for col in cd_train.select_dtypes(['object']).columns:
    print(col,':',cd_train[col].nunique())

Product : 12
Sub-product : 47
Issue : 95
Sub-issue : 68
Consumer complaint narrative : 74019
Company public response : 10
Company : 3276
State : 62
ZIP code : 25962
Tags : 3
Consumer consent provided? : 4
Submitted via : 6
Company response to consumer : 7
Timely response? : 2
Consumer disputed? : 2


In [8]:
cd_train.isnull().sum()

Product                              0
Sub-product                     138473
Issue                                0
Sub-issue                       292625
Consumer complaint narrative    403327
Company public response         388029
Company                              0
State                             3839
ZIP code                          3848
Tags                            411215
Consumer consent provided?      342934
Submitted via                        0
Company response to consumer         0
Timely response?                     0
Consumer disputed?                   0
Complaint ID                         0
day_diff                             0
dtype: int64

In [9]:
len(pd.isnull(cd_train['Tags']))
len(cd_train)

478421

In [10]:
for col in ['Sub-product','Sub-issue','Consumer complaint narrative',
            'Company public response','Tags','Consumer consent provided?']:
    varname=col.replace('-','_').replace('?','').replace(" ",'_')+'_isNan'
    cd_train[varname]=np.where(pd.isnull(cd_train[col]),1,0)
    cd_train.drop([col],1,inplace=True)
    cd_test[varname]=np.where(pd.isnull(cd_test[col]),1,0)
    cd_test.drop([col],1,inplace=True)
    

In [11]:
cd_train.head(4)

Product                                     Issue  \
0              Credit card                         Billing statement   
1  Bank account or service  Making/receiving payments, sending money   
2         Credit reporting    Incorrect information on credit report   
3              Credit card                         Billing statement   

                 Company State ZIP code Submitted via  \
0  Wells Fargo & Company    MI    48342           Web   
1      Santander Bank US    PA    18042      Referral   
2                Equifax    CA    92427      Referral   
3           U.S. Bancorp    GA    305XX           Web   

      Company response to consumer Timely response? Consumer disputed?  \
0          Closed with explanation              Yes                 No   
1                           Closed              Yes                 No   
2  Closed with non-monetary relief              Yes                 No   
3      Closed with monetary relief              Yes                 No   

   Complaint ID          day_diff  Sub_product_isNan  Sub_issue_isNan  \
0        856103    86400000000000                  1                1   
1       1034666   518400000000000                  0                1   
2        756363  1814400000000000                  1                0   
3       1474177                 0                  1                1   

   Consumer_complaint_narrative_isNan  Company_public_response_isNan  \
0                                   1                              1   
1                                   1                              1   
2                                   1                              1   
3                                   0                              0   

   Tags_isNan  Consumer_consent_provided_isNan  
0           0                                1  
1           1                                1  
2           1                                1  
3           0                                0

In [12]:
for col in cd_train.select_dtypes(['object']).columns:
    print(col,':',cd_train[col].nunique())

Product : 12
Issue : 95
Company : 3276
State : 62
ZIP code : 25962
Submitted via : 6
Company response to consumer : 7
Timely response? : 2
Consumer disputed? : 2


In [13]:
for col in ['ZIP code','Company']:
    cd_train.drop([col],axis=1,inplace=True)
    cd_test.drop([col],axis=1,inplace=True)

In [14]:
#cd_train['Consumer disputed?']=np.where(cd_train['Consumer disputed?']=="Yes",1,0)

In [15]:
k=cd_train['Issue'].value_counts()
for val in k.axes[0][0:10]:
    varname='Issue_'+val.replace(',','_').replace(' ','_')
    cd_train[varname]=np.where(cd_train['Issue']==val,1,0)
    cd_test[varname]=np.where(cd_test['Issue']==val,1,0)
del cd_train['Issue']
del cd_test['Issue']

In [16]:
for col in cd_train.select_dtypes(['object']).columns:
    print(col,':',cd_train[col].nunique())

Product : 12
State : 62
Submitted via : 6
Company response to consumer : 7
Timely response? : 2
Consumer disputed? : 2


In [17]:
k=cd_train['State'].value_counts()
for val in k.axes[0][0:15]:
    varname='State_'+val.replace(',','_').replace(' ','_')
    cd_train[varname]=np.where(cd_train['State']==val,1,0)
    cd_test[varname]=np.where(cd_test['State']==val,1,0)
del cd_train['State']
del cd_test['State']

In [18]:
for col in cd_test.columns: 
    print(col)

Product
Submitted via
Company response to consumer
Timely response?
Complaint ID
day_diff
Sub_product_isNan
Sub_issue_isNan
Consumer_complaint_narrative_isNan
Company_public_response_isNan
Tags_isNan
Consumer_consent_provided_isNan
Issue_Loan_modification_collection_foreclosure
Issue_Incorrect_information_on_credit_report
Issue_Loan_servicing__payments__escrow_account
Issue_Cont'd_attempts_collect_debt_not_owed
Issue_Account_opening__closing__or_management
Issue_Disclosure_verification_of_debt
Issue_Communication_tactics
Issue_Deposits_and_withdrawals
Issue_Application__originator__mortgage_broker
Issue_Billing_disputes
State_CA
State_FL
State_TX
State_NY
State_GA
State_NJ
State_PA
State_IL
State_VA
State_MD
State_OH
State_NC
State_MI
State_AZ
State_WA


In [19]:
col=['Product','Submitted via','Company response to consumer']
cd_train = pd.get_dummies(data=cd_train,columns = col,drop_first=True)

cd_test = pd.get_dummies(data=cd_test,columns = col,drop_first=True)


In [20]:
#Initialize the labelEncoder

from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
cd_train['Consumer disputed?'] = le.fit_transform(cd_train['Consumer disputed?'])
cd_train['Timely response?'] = le.fit_transform(cd_train['Timely response?'])

In [21]:
#Initialize the labelEncoder

from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
cd_test['Timely response?'] = le.fit_transform(cd_test['Timely response?'])

In [22]:
x=cd_train.drop(['Consumer disputed?','Complaint ID'],1)
y=cd_train['Consumer disputed?']

# Logistic Regression

In [23]:
from sklearn.linear_model import LogisticRegression

In [24]:
clf=LogisticRegression()

In [25]:
clf.fit(x,y)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [26]:
# Prediction
prediction=clf.predict(x)

In [27]:
#Import scikit-learn metrics module for accuracy calculation
from sklearn import metrics
# Model Accuracy, how often is the classifier correct?
print("Accuracy:",metrics.accuracy_score(y,prediction))

Accuracy: 0.7814895249163394


In [28]:
from sklearn.metrics import roc_curve, roc_auc_score
roc_auc_score(y, clf.predict(x))

0.5002183614410464

# Random Forest Classifier

In [29]:
#Import Random Forest Model
from sklearn.ensemble import RandomForestClassifier

In [30]:
#Create a Gaussian Classifier
clf2=RandomForestClassifier(n_estimators=100)

In [31]:
#Train the model using the training sets y_pred=clf.predict(X_test)
clf2.fit(x,y)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=None, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [32]:
y_pred=clf2.predict(x)

In [33]:
#Import scikit-learn metrics module for accuracy calculation
from sklearn import metrics
# Model Accuracy, how often is the classifier correct?
print("Accuracy:",metrics.accuracy_score(y,prediction))

Accuracy: 0.7814895249163394


In [34]:
from sklearn.metrics import roc_curve, roc_auc_score
roc_auc_score(y, clf2.predict(x))

0.5819283934768472

In [35]:
prediction=np.where(clf2.predict(cd_test.drop(['Complaint ID'],1))==1,"Yes","No")
submission=pd.DataFrame(list(zip(cd_test['Complaint ID'],list(prediction))),
                       columns=['Complaint ID','Consumer disputed?'])

In [36]:
submission.head(4)

Complaint ID Consumer disputed?
0        675956                 No
1       1858795                 No
2         32637                 No
3       1731374                 No

In [37]:
submission.to_csv('sample_submission.csv',index=False)

This submission will get you auc score of approx 0.50, slightly less than whats required for passing the course. You'll have to make changes 

# Finally roc auc score is achieved 0.58 from Random Forest Classifier